pm_kext/col/verdict_cache.c

```diff
@@ -135,34 +156,55 @@ int teardownVerdictCache(VerdictCache *verdictCache) {
  * @return error code
  *
  */
-int addVerdict(VerdictCache *verdictCache, PortmasterPacketInfo *packetInfo, verdict_t verdict) {
-    if (!verdictCache || !packetInfo || !verdict) {
-        ERR("add_verdict NULL pointer exception verdictCache=0p%Xp, packetInfo=0p%Xp, verdict=0p%Xp ", verdictCache, packetInfo, verdict);
+int addVerdict(VerdictCache *verdictCache, PortmasterPacketInfo *packetInfo, verdict_t verdict, PortmasterPacketInfo **removedPacketInfo) {
+    if (verdictCache == NULL || packetInfo == NULL || verdict == 0) {
+        ERR("addVerdict NULL pointer exception verdictCache=0p%Xp, packetInfo=0p%Xp, verdict=0p%Xp ", verdictCache, packetInfo, verdict);
         return 1;
     }
 
-    VerdictCacheItem *newItem = _ALLOC(sizeof(VerdictCacheItem), 1);
-    if(!newItem) {
-        ERR("add_verdict tried to add NULL-Pointer verdict");
-        return 2;
-    }
+    counter++;
 
-    newItem->packetInfo = packetInfo;
-    newItem->verdict = verdict;
+    VerdictCacheItem *newItem = NULL;
 
-    // insert as first item
-    if (verdictCache->head != NULL) {
-        newItem->next = verdictCache->head;
-        verdictCache->head->prev = newItem;
+    VerdictCacheKey key = getCacheKey(packetInfo);
+    #pragma warning(suppress : 4127) //  warning C4127: conditional expression is constant -> if generated by macro
+    HASH_FIND(hh, verdictCache->map, &key, sizeof(VerdictCacheKey), newItem);
+    if(newItem != NULL) {
+        INFO("addVerdict packet was already in");
```

Comment:
If I understand correctly, this aborts adding the verdict to the cache when it already exists? Is there a way we can update the verdict?